File: src/main/code/y2024/java/Day1.java
package y2024.java;

import day.Day;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Day1 extends Day {
    public Day1() {
        super(2024, 1);
    }

    public List<List<Long>> parseInput() {
        List<Long> left = new ArrayList<>();
        List<Long> right = new ArrayList<>();
        List<String> lines = getInput().lines().toList();
        for (String line : lines) {
            String[] parts = line.split("\\s+");
            left.add(Long.parseLong(parts[0]));
            right.add(Long.parseLong(parts[1]));
        }
        return List.of(left, right);
    }

    @NotNull
    @Override
    protected Object part1() {
        List<List<Long>> lists = parseInput();
        List<Long> left = lists.get(0);
        List<Long> right = lists.get(1);

        left.sort(Long::compareTo);
        right.sort(Long::compareTo);

        assert left.size() == right.size();

        long sum = 0;
        for (int i = 0; i < left.size(); i++) {
            sum += Math.abs(left.get(i) - right.get(i));
        }
        return sum;
    }

    @NotNull
    @Override
    protected Object part2() {
        List<List<Long>> lists = parseInput();
        List<Long> left = lists.get(0);
        List<Long> right = lists.get(1);

        Map<Long, Long> frequency = new HashMap<>();
        for (Long l : right) {
            frequency.put(l, frequency.getOrDefault(l, 0L) + 1);
        }
        return left.stream().map(l -> l * frequency.getOrDefault(l, 0L)).reduce(0L, Long::sum);
    }

    public static void main(String[] args) {
        new Day1().run();
    }
}

File: src/main/code/y2024/kotlin/Day1.kt
package y2024.kotlin

import day.Day
import kotlin.math.absoluteValue

class Day1: Day(2024, 1) {

    private fun parseInput(): Pair<List<Int>, List<Int>> {
        val ws = Regex("\\s+")
        return input.split("\n").map { it.split(ws) }.map { it[0].toInt() to it[1].toInt() }.unzip()
    }

    override fun part1(): Int {
        val (l, r) = parseInput()
        return l.sorted().zip(r.sorted()).sumOf { (it.first - it.second).absoluteValue }
    }

    override fun part2(): Int {
        val (l, r) = parseInput()
        val freqs = HashMap<Int, Int>()
        r.forEach { freqs[it] = freqs.getOrDefault(it, 0) + 1 }
        return l.sumOf { it * freqs.getOrDefault(it, 0) }
    }
}

fun main() {
    Day1().run()
}